### PR TITLE
Card Template, Steven please look

### PR DESCRIPTION
**Pull request by srmcconomy**

### CODE REVIEW

File: src/shared/cards/Card Template.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  async onPlay(player) {
+    // Vanilla Effects
+    player.draw(1);
+    player.actions++;
+    player.buys++;
+    player.money++;
+
+    // Trash card(s) in Hand, is it game.currentPlayer, or just Player?
+    const cards = await game.currentPlayer.selectCards(min, max, predicate);
+    for card in cards {
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js]:
```
cards.forEach(card => {
  player.trash(card);
});
```

File: src/shared/cards/Card Template.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    // Draw card and do something if()
+    card = player.deck.last();
+    if (card.types.has('Victory')) {  // look at top card in deck, discard if it is a victory card
+      player.moveCard(this, player.deck, player.discard);
+    }
+
+    // Draw multiple cards and do something if(), basically emulating library
+    const cards = player.setAside(min, max);
+    for card in cards {
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js]: `forEach`

File: src/shared/cards/Card Template.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      if card.types.has('Action') {
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js]: `if` always needs brackets

File: src/shared/cards/Card Template.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        await choice = player.chooseOption(['hand', 'discard']);
+        if (choice == 'hard') {
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js]: always `===` not `==`

File: src/shared/cards/Card Template.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+          player.moveCard(card, player.setAside, player.hand);
+        }
+        else {player.moveCard(card, player.setAside, player.discard);}
+      }
+    }
+
+    // Oponents draw card(s) and you do something , basically thief
+    await game.forEachOtherPlayer(async player => {
+      cards = player.deck[end-x:end];
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js]: thats not a thing

Comment (srmcconomy) [reply src/shared/cards/Card Template.js]: `player.deck.spliceLast(x)`
see `Pile.js`

Comment (srmcconomy) [reply src/shared/cards/Card Template.js]: which actually modifies deck

File: src/shared/cards/Card Template.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    // Other players discard down to X
+    await game.forEachOtherPlayer(async player => {
+      if (player.hand.size > X) {
+        const cards = await player.selectCards(player.hand.count - X, 'Select card(s) to discard');
+        cards.forEach(card => player.discard(card));
+      }
+    });
+
+    // Gain a Card
+    const supply = await player.chooseSupplyWhere(supply => (supply.last().cost <= 5 && supply.last.types.has('Treasure')));
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js]: `supply.last().types.has('Treasure')`

File: src/shared/cards/Card Template.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    player.gain(supply.last().name);
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js]: just `supply.last()`

File: src/shared/cards/Card Template.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
+import Card from 'cards/Card';
+import { Set } from 'immutable';
+
+export default class CardName extends Card {
+  name = 'Card Name';
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js]: because of a weird thing in JS this can't be called `name` so I changed it to `title`

Comment (srmcconomy) [reply src/shared/cards/Card Template.js]: also all of these should be `static`
```
static title = 'Card Name';

File: src/shared/cards/Card Template.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    card = player.deck.last();
+    if (card.types.has('Victory')) {  // look at top card in deck, discard if it is a victory card
+      player.moveCard(this, player.deck, player.discard);
+    }
+
+    // Draw multiple cards and do something if(), basically emulating library
+    const cards = player.setAside(min, max);
+    for card in cards {
+      if card.types.has('Action') {
+        await choice = player.chooseOption(['hand', 'discard']);
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js]:
```
const choice = await player.chooseOption(['hand', 'discard']);

File: src/shared/cards/Card Template.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        if (choice == 'hard') {
+          player.moveCard(card, player.setAside, player.hand);
+        }
+        else {player.moveCard(card, player.setAside, player.discard);}
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js]: the standard is to put this `else` on the same line as `}` above it, and but the body of the else on a new line
```
} else {
  player.moveCard(...);
}
```

File: src/shared/cards/Card Template.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
+import Card from 'cards/Card';
+import { Set } from 'immutable';
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js]: Okay this is confusing but
`Set` is a built in native thing in JS so you don't need to import it
`Immutable.Set` is a different thing that I'm not using server-side any more, so just don't have this line basically

File: src/shared/cards/Card Template.js (hunk 4 above)
Comment (srmcconomy) [reply src/shared/cards/Card Template.js]: won't make a difference but its the standard... 
`===` compares type and value so `1 === '1'` is false
`==` only compares value so `1 == '1'` is true

File: src/shared/cards/Card Template.js (hunk 1 above)
Comment (generalpepper) [reply src/shared/cards/Card Template.js]: Is selectCards called correctly via game.currentPlayer rather than just player?

Comment (srmcconomy) [reply src/shared/cards/Card Template.js]: actually `game` is not defined, you want `player.game`
and `player` is the player that played the card while `player.game.currentPlayer` is the player whose turn it is
there would only be a difference if you play a card when its not your turn somehow

File: src/shared/cards/Card Template.js (hunk 5 above)
Comment (generalpepper) [reply src/shared/cards/Card Template.js]: Am i redefining player in here?
player.draw(1) refers to the current player i think right?
So anything with other players should have a different variable name

**Comment by generalpepper** [comment]
Different question;
in selectCards() what happens if there aren't enough to meet the requirement?

```js
    async selectCards(min, max, predicate, from = 'hand') {
      console.log('select-cards');
      const filteredCards = this[from].filter(predicate);
      if (filteredCards.size === 0) {
        return [];
      }
      const payload = {
        type: 'select-cards',
        min,
        max,
        cards: filteredCards.toIDArray(),
      };
      const ids = await this.getInputAndNotifyDirty(payload, arr => arr.every(id => filteredCards.hasID(id)));
      return ids.map(id => Model.fromID(id));
    }
```

Should there be something like
```js
      if (filteredCards.size === 0) {
        return [];
      } else if (filteredCards.size < max) {
        max = filteredCards.size;
      }
```

File: src/shared/cards/Card Template.js (hunk 5 above)
Comment (generalpepper) [reply src/shared/cards/Card Template.js]: Probably don't want to actually use splice, just move cards to set aside location and then do stuff

**Comment by generalpepper** [comment]
can we have a setAside function that moves x cards from the top of a deck to a temporary location

and

some other function that moves the set aside cards to the discard pile (after we have taken out the ones we want and put them wherever.

File: src/shared/cards/Card Template.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+export default class CardName extends Card {
+  static title = 'Card Name';
+  cost = 4;
+  VP = 3;
+  coin = 2;
+  types = new Set(['Action', 'Treasure', 'Attack', 'Reaction', 'Victory']);
+  async onPlay(player) {
+    // Vanilla Effects
+    player.draw(1);
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js:11]: I think you'll have to do `await player.draw(1)`
same for `player.trash`, `player.play`, `player.discard`, etc

File: src/shared/cards/Card Template.js [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    player.trash(this);
+
+    // Draw card and do something if()
+    card = player.deck.last();
+    if (card.types.has('Victory')) {  // look at top card in deck, discard if it is a victory card
+      player.moveCard(this, player.deck, player.discard);
+    }
+
+    // Draw multiple cards and do something if(), basically emulating library
+    const cards = player.setAside(min, max);
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js:32]: I made `player.draw` return the cards that you draw
I also made `player.draw` take a parameter that lets you decide where the card goes
eg
```
while (player.hand.size < 7) {
  await player.draw(1, async card => {
    if (card.types.has('Action') {
      const choice = await player.selectOption(['Put in hand', 'Set aside']);
      return [player.hand, player.aside][choice];
    }
    return player.hand;
  });
}
await player.moveCards(player.aside, player.discardPile, player.aside.size);
```

File: src/shared/cards/Card Template.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        }
+      }
+    });
+    player.unsetAside();
+
+
+    // Oponents draw card(s) and you do something , basically thief
+    await game.forEachOtherPlayer(async otherPlayers => {
+      const cards = otherPlayers.setAside(x,x);
+      const card = await otherPlayers.selectCards(1, 1, card => card.types.has(['Treasure'], cards);
```
Comment (srmcconomy) [inline src/shared/cards/Card Template.js:49]: you can make that decision for them a lot of the time
```
if (cards.size > 1 && cards.every(card => card.types.has('Treasure')) {
  const card = await otherPlayer.selectCards(1, 1, cards);
  await otherPlayer.trash(card);
  await otherPlayer.discard(cards.find(c => c !== card));
} else {
  for (let i = 0; i < cards.length; i++) {
    const card = cards.get(i);
    if (card.types.has('Treasure')) {
      await otherPlayer.trash(card);
    } else {
      await otherPlayer.discard(card);
    }
  }
}
```

Comment (srmcconomy) [reply src/shared/cards/Card Template.js:49]: You also don't need to use `aside` unless the card actually says put aside like in Library

**Pull request closed without merging**